src/app/services/file-synchronizer.service.ts

```diff
@@ -0,0 +1,140 @@
+            break;
+          default:
+            console.log(response, filePath, dbType);
+        }
+      });
+    });
+  }
+
+  private upsertSonggroup(filePath: string, data: Songgroup): Promise<Songgroup> {
+    return this.apiService.generateFileCreateRequest(filePath, data).then(res => {
+      return new Promise<Songgroup>(resolve => resolve(data));
+    });
+  }
+
+  private upsertSong(filePath: string, data: Song): Promise<Song> {
+    return this.apiService.generateFileUpdateRequest(filePath, data).then( res => {
+      console.log('upsert file', res);
+      switch (res.status) {
+        case 201:
+          return new Promise<Song>(resolve => resolve(data));
+        case 300:
+          console.log(res.payload);
+          return this.mergeService.mergeSong(res.payload.indexedVersion, res.payload.currentVersion, data).then(song => {
+            return this.upsertSong(filePath, <Song>song);
+          });
+        case 404:
+          return this.apiService.generateFileCreateRequest(filePath, data).then(response => {
+            return new Promise<Song>(resolve => resolve(data));
+          });
+      }
+    });
+  }
+
+  public getSongs(): Promise<Song[]> {
+    return this.dexieService.getAll(DATABASES.songs).then(res => {
+      return this.map<Song>(res);
+    });
+  }
+
+  public getSong(songid: string): Promise<Song> {
+    return this.dexieService.getByKey(DATABASES.songs, songid).then(res => {
+      return <Song>res;
+    });
+  }
+
+  public saveSong(song: Song): Promise<Song> {
+    return this.upsertSong(path.join(DATABASES.songs, song.id + '.song'), song).then(s => {
+      return this.dexieService.upsert(DATABASES.songs, s).then(() => {
+        return new Promise<Song>(resolve => resolve(<Song>s));
+      });
+    });
+  }
+
+  public deleteSong(songid: string) {
+    return this.dexieService.delete(DATABASES.songs, songid).then(() => {
+      this.apiService.generateDeleteFileRequest(path.join(DATABASES.songs, songid + '.song'));
+    });
+  }
+
+  public getSonggroups(): Promise<Songgroup[]> {
+    return this.dexieService.getAll(DATABASES.songgroups).then(res => {
+      return this.map<Songgroup>(res);
+    });
+  }
+
+  public getSonggroup(songgroupid: string): Promise<Songgroup> {
+    return this.dexieService.getByKey(DATABASES.songgroups, songgroupid).then(res => {
+      return <Songgroup>res;
+    });
+  }
+
+  public saveSonggroup(songgroup: Songgroup): Promise<Songgroup> {
+    return this.dexieService.upsert(DATABASES.songgroups, songgroup).then(res => {
+      return this.upsertSonggroup(path.join(DATABASES.songgroups, songgroup.id + '.songgroup'), songgroup);
+    });
+  }
+
+  public deleteSonggroup(songgroupid: string) {
+    return this.dexieService.delete(DATABASES.songgroups, songgroupid).then(() => {
+      this.apiService.generateDeleteFileRequest(path.join(DATABASES.songgroups, songgroupid + '.songgroup'));
+    });
+  }
+
+  private map<T>(array) {
```

Comment:
you can shorten this if you want
```suggestion
  private map<T>(array): Array<T> { return (array || []).map(v => <T>v) }
```

Reply:
maybe the naming should also be more like 'castMap' or something similar